src/formats/png/reader.zig

```diff
@@ -0,0 +1,1041 @@
+    const idat = "IDAT";
+    const idat_id = asU32(idat);
+
+    fn init(
+        stream: *Image.Stream,
+        processors: []ReaderProcessor,
+        chunk_process_data: *ChunkProcessData,
+    ) Self {
+        var crc = Crc32.init();
+        crc.update(idat);
+        return .{
+            .stream = stream,
+            .processors = processors,
+            .chunk_process_data = chunk_process_data,
+            .remaining_chunk_length = chunk_process_data.chunk_length,
+            .crc = crc,
+        };
+    }
+
+    fn fillBuffer(self: *Self, to_read: usize) Image.ReadError!usize {
+        std.mem.copy(u8, self.buffer[0 .. self.end - self.pos], self.buffer[self.pos..self.end]);
+        self.pos = self.end - self.pos;
+        var max = self.buffer.len;
+        if (max - self.pos > self.remaining_chunk_length) max = self.pos + self.remaining_chunk_length;
+        const len = try self.stream.read(self.buffer[self.pos..max]);
+        self.end = self.pos + len;
+        self.crc.update(self.buffer[self.pos..self.end]);
+        return if (len < to_read) len else to_read;
+    }
+
+    fn read(self: *Self, dest: []u8) Image.ReadError!usize {
+        if (self.remaining_chunk_length == 0) return 0;
+        const new_dest = dest;
+
+        var reader = self.stream.reader();
+        var to_read = new_dest.len;
+        if (to_read > self.remaining_chunk_length) to_read = self.remaining_chunk_length;
+        if (to_read > self.end - self.pos) to_read = try self.fillBuffer(to_read);
+        std.mem.copy(u8, new_dest[0..to_read], self.buffer[self.pos .. self.pos + to_read]);
+        self.remaining_chunk_length -= @intCast(u32, to_read);
+        self.pos += to_read;
+
+        if (self.remaining_chunk_length == 0) {
+            // First read and check CRC of just finished chunk
+            const expected_crc = try reader.readIntBig(u32);
+            if (self.crc.final() != expected_crc) return Image.ReadError.InvalidData;
+
+            try callChunkProcessors(self.processors, self.chunk_process_data);
+
+            self.crc = Crc32.init();
+            self.crc.update(idat);
+
+            // Try to load the next IDAT chunk
+            const chunk = try utils.readStructBig(reader, png.ChunkHeader);
+            if (chunk.type == idat_id) {
+                self.remaining_chunk_length = chunk.length;
+            } else {
+                // Return to the start of the next chunk so code in main struct can read it
+                try self.stream.seekBy(-@sizeOf(png.ChunkHeader));
+            }
+        }
+
+        return to_read;
+    }
+};
+
+const IDATReader = std.io.Reader(*IDatChunksReader, Image.ReadError, IDatChunksReader.read);
+
+pub fn loadHeader(stream: *Image.Stream) Image.ReadError!png.HeaderData {
+    var reader = stream.reader();
+    var sig: [png.magic_header.len]u8 = undefined;
+    try reader.readNoEof(sig[0..]);
+    if (!mem.eql(u8, sig[0..], png.magic_header)) return Image.ReadError.InvalidData;
```

Comment:
I see that you have a lot of one-liner `if` with no brackets. In my experience, not using {} was always error prone.